src/core/ext/filters/client_channel/subchannel.cc

```diff
@@ -777,12 +766,16 @@ void Subchannel::Orphan() {
     subchannel_pool_->UnregisterSubchannel(key_, this);
     subchannel_pool_.reset();
   }
-  MutexLock lock(&mu_);
-  GPR_ASSERT(!shutdown_);
-  shutdown_ = true;
-  connector_.reset();
-  connected_subchannel_.reset();
-  health_watcher_map_.ShutdownLocked();
+  {
+    MutexLock lock(&mu_);
+    GPR_ASSERT(!shutdown_);
+    shutdown_ = true;
+    connector_.reset();
+    connected_subchannel_.reset();
+    health_watcher_map_.ShutdownLocked();
+  }
+  // Drain any connectivity state notifications after releasing the mutex.
+  work_serializer_.DrainQueue();
```

Comment:
On line 784 and line 790 we lock the mutex but we don't drain the work serializer. What is the chance that `get_or_add` will in the future do a WorkSerializer schedule and cause things to get stuck, similar for line 790